examples/scripts/interactive/compvis-viewer.js

```diff
@@ -0,0 +1,603 @@
+  }
+
+  // get colors
+  var startRed = parseInt(startColor.substr(0, 2), 16)
+  var startGreen = parseInt(startColor.substr(2, 2), 16)
+  var startBlue = parseInt(startColor.substr(4, 2), 16)
+
+  var endRed = parseInt(endColor.substr(0, 2), 16)
+  var endGreen = parseInt(endColor.substr(2, 2), 16)
+  var endBlue = parseInt(endColor.substr(4, 2), 16)
+
+  // calculate new color
+  var diffRed = endRed - startRed
+  var diffGreen = endGreen - startGreen
+  var diffBlue = endBlue - startBlue
+
+  diffRed = ((diffRed * percent) + startRed).toString(16).split('.')[0]
+  diffGreen = ((diffGreen * percent) + startGreen).toString(16).split('.')[0]
+  diffBlue = ((diffBlue * percent) + startBlue).toString(16).split('.')[0]
+
+  // ensure 2 digits by color (necessary)
+  if (diffRed.length === 1) diffRed = '0' + diffRed
+  if (diffGreen.length === 1) diffGreen = '0' + diffGreen
+  if (diffBlue.length === 1) diffBlue = '0' + diffBlue
+
+  return '0x' + diffRed + diffGreen + diffBlue
+}
+
+function makeGradientArray () {
+  var gradientArray = []
+  for (var count = 0; count < 101; count++) {
+    var newColor = getGradientColor('FF0000', 'FFFFFF', '0000FF', (0.02 * count))
+    var numColor = parseInt(Number(newColor), 10)
+    gradientArray.push(numColor)
+  }
+  return gradientArray
+}
+
+var gradientArray = makeGradientArray()
+
+// var ligandSele = '( not polymer or not ( protein or nucleic ) ) and not ( water or ACE or NH2 )'
+
+var pocketRadius = 0
+var pocketRadiusClipFactor = 1
+
+var cartoonRepr, neighborRepr, ligandRepr, contactRepr, pocketRepr, labelRepr, customRepr
+
+var heatMap, customPercent
+
+var struc
+var csv
+var neighborSele
+var sidechainAttached = false
+
+const csvResNumCol = 4
+const csvWtProbCol = 7
+const csvPrAaCol = 6
+const csvPrProbCol = 8
+
+var firstResNum
+
+function loadStructure (proteinFile, csvFile) {
+  struc = undefined
+  stage.setFocus(0)
+  stage.removeAllComponents()
+  clipNearRange.value = 0
+  clipRadiusRange.value = 100
+  pocketOpacityRange.value = 0
+  cartoonCheckbox.checked = true
+  customCheckbox.checked = false
+  hydrophobicCheckbox.checked = false
+  hydrogenBondCheckbox.checked = true
+  weakHydrogenBondCheckbox.checked = false
+  waterHydrogenBondCheckbox.checked = true
+  backboneHydrogenBondCheckbox.checked = true
+  halogenBondCheckbox.checked = true
+  metalInteractionCheckbox.checked = true
+  saltBridgeCheckbox.checked = true
+  cationPiCheckbox.checked = true
+  piStackingCheckbox.checked = true
+  return Promise.all([
+    stage.loadFile(proteinFile, { defaultRepresentation: true }),
+    NGL.autoLoad(csvFile, {
+      ext: 'csv',
+      delimiter: ' ',
+      comment: '#',
+      columnNames: true
+    })
+  ]).then(function (ol) {
+    struc = ol[0]
+    csv = ol[1].data
+
+    firstResNum = parseInt(csv[0][csvResNumCol])
+
+    heatMap = NGL.ColormakerRegistry.addScheme(function (params) {
+      this.atomColor = function (atom) {
+        for (var i = 0; i < csv.length; i++) {
```

Comment:
Then here: 
```javascript
const csvRow = residueData[atom.resno]
const wtProb = parseFloat(csvRow[csvWtProbCol])

// Etc...
```

Reply:
I am able to get all of these suggestions working on some proteins, but not on 2isk and vice-versa. the problem is that the csv data does not always match up with the structure data. For instance I was using atom.residueIndex as the iterator ( instead of atom.resno -- I changed to residueData accordingly), but atom.residueIndex is 219 long while the csv is 218 long. I tried your way, with resNum and resno, and had same issue on a different protein.

If the csv and structure data do not match does this require me to use csv.length as the limit to iteration (i.e. for loop in atomColor) to avoid getting "Undefined is not an object" error? Or is there another way that I'm missing?

I can use qmean example to simplify color generation, but with a for loop.

I'll try to have an updated example where you can switch proteins easily pushed by tomorrow

Reply:
See my comment in your updated version - I think you just need to check cases where residueData[atom.resno] is undefined and return a default color.